src/H5Tconv.c

```diff
@@ -3801,6 +3818,57 @@ H5T__conv_vlen(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata, const H5T
     } /* end switch */
 
 done:
+    /* Release converted elements on error */
+    if (ret_value < 0 && conversions_made) {
+        size_t dest_count;
+
+        /* Set up for first pass to destroy references */
+        if (nelmts < orig_nelmts || (convert_forward && elmtno < safe)) {
+            dest_count = orig_nelmts - nelmts;
+
+            /* Set pointer to correct location, based on direction chosen */
+            if (convert_forward) {
+                d = (uint8_t *)buf;
+                dest_count += elmtno; /* Include partial iteration in first pass, for forward conversions */
+            }
+            else
+                d = (uint8_t *)buf + (nelmts * orig_d_stride);
+
+            /* Destroy references that have already been converted */
+            while (dest_count > 0) {
+                H5_GCC_DIAG_OFF("discarded-qualifiers")
+                H5_CLANG_DIAG_OFF("incompatible-pointer-types-discards-qualifiers")
+                H5T_vlen_reclaim_elmt(d, dst); /* Ignore errors at this point */
```

Comment:
Rather than hiding the const warnings here due to `H5T_vlen_reclaim_elmt` taking a `H5T_t *` parameter, I think we can just update `H5T_vlen_reclaim_elmt` to take a `const H5T_t *` instead. All it does is pass that parameter to `H5T__vlen_reclaim`, which takes a `const H5T_t *`.

Reply:
Fixed